---
title: "Testing DESEQ2"
output: html_notebook
---

Iteration of DESEQ2 over: Region level, then treatment comparison

Region level - e.g. Tissue region, indicates all DE data from same tissue and experiment. These are analysed independently from every other region.

Within a region, the treatments are compared.


However, for downstream plotting, the normalised data from all regions is calculated. Particularly important for co-expression related analysis.


To normalise gene expression across all tissues, the 'DESEQ2 - all tissues - FULL - **' needs to be used.

#Step 1a in pipeline

```{r echo=F, warning = FALSE, message = FALSE}
library(plyr)
library(DESeq2)
library(data.table)
library(tidyverse)
library(RColorBrewer)
library(genefilter)
library(pheatmap)
library(gplots)
library(ggrepel)
```


# 1. Import metadata
```{r}

#import data about columns, that is each column should be metadata about the sample/animal. Particularly include all treatment info and animal ID in columns. 
coldata <-  data.table::fread(file = "../Sheep_RNAseq/Inputs/coldata_new_names_20201105.csv")


format_data <- function(colData, counts_matrix, column_containing_rownames = "rowname") {
  #check that coldata rownames match columns of count_data
  
  
  
  count_data1 <- count_data %>% dplyr::select(gene_ensembl, coldata$rowname)
}




```




# 2. import all count tables and join
```{r matrix_input, echo=T}
#read in count table - hypothalamus
getwd()
cts_hypo_liv <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_hypothalamus-liver.txt", header = T)

colnames(cts_hypo_liv)[1] <- "gene_ensembl"

#rename incorrect entries:
cts_hypo_liv <- cts_hypo_liv %>% 
  dplyr::rename("VMH_21_HCP-HP-UMEI" = "VMH_21_HCP-HP-RMEI", 
                "VMH_02_HCP-LP-UMEI" = "VMH_02_LCP-LP-UMEI")


#rename LAT to LHA
colnames(cts_hypo_liv) <- stringr::str_replace_all(string = names(cts_hypo_liv), pattern = "LAT", replacement = "LHA")


#read in count table - GIT & ST
cts_git_st <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_matrix_DUO_RUM_ST_20200901.tabular")

colnames(cts_git_st)[1] <- "gene_ensembl"

#Join cts together

cts_all <- dplyr::full_join(cts_hypo_liv, cts_git_st)


###### ************
# Use new function
cts_filtered <- DE4Rumi::check_count_matrix(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = rowname
                            )

```


```{r}


#order the columns in cts to match coldata info (also removes annotation data) *** also removes teh +bST data
cts <- dplyr::select(cts_all, gene_ensembl, coldata$rowname)
#names(cts)


#rename colnames???
#colnames(cts) <- c("gene_ensembl",coldata$rowname0)


#set cts with rownames for dds creation                                     
cts0 <- cts %>% tibble::column_to_rownames(var = "gene_ensembl") %>% as.data.frame()

#coldata1 <- coldata %>%  column_to_rownames(var = "rowname0")
```




# 2. convert ensemble ID to gene ID

```{r}
# library(gprofiler2)
# #this connects to the online server for gProfiler - so takes some time
# converted0 <- gprofiler2::gconvert(query = cts_all$gene_ensembl,
#          organism = "oaries",
#          target = "ENSG", # or ENSG | AFFY_BOVINE
#          filter_na = T,
#          mthreshold = 1)
# 
# converted_genes <- converted0 %>% dplyr::select(input, name, description)
# 
# cts_a <- left_join(cts_all,converted_genes, by = c("gene_ensembl" = "input"))
# 
# ### important - this produces duplicate gene names from annotation - however gene_ensembl remain unique
# 
# paste("duplicate genes in annotation: ",nrow(cts_a[duplicated(cts_a$name),]))


test_annot <- DE4Rumi::annotate_gene_ensembl(cts_all, organism = "oaries")
```




*******************************************************************************************************************************

## ~Set region and hormone selection for individual DESEQ
Run code chunks below this for subsequant analysis of individual regions, as above imported data and annotation need only be run once per session.
```{r input_parameters, echo=T}
#set region for downstream formulas - as character


region <- "" #options: ARC, LHA, VMH, LIV /// RUM, DUO, ST



#note: need to write different model design and results contrasts if hormone comparison wants to be tested, however this can test diet treatments in the +bST dataset
```

*******************************************************************************************************************************



#Test summarized experiment
https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html

All data needs to be ordered to match, and have rownames
To call 
```{r}


#converted_genes0 <- converted_genes %>% column_to_rownames(var = "input")


test_data <- 
SummarizedExperiment::SummarizedExperiment(assays = cts %>% tibble::column_to_rownames(var = "gene_ensembl") %>% as.matrix(),
                     rowData = test_annot, #this does not need rownames, but if they are set they become the new rownames
                     colData = coldata, #this does not need rownames, but if they are present they become new colnames
                     metadata = list(author = "David Innes"))


# test_data@colData
# test_data$Diet
# 
# #getters
# assay(test_data)
# rowData(test_data)
# colData(test_data)
# metadata(test_data)
# 
# #subset
# test_data[1:5, 1:3] %>% assay()

#subset by category
test_data_2 <-  test_data[,test_data$Region == "LIV"]




#create DESEQ object
dds <- DESeq2::DESeqDataSet(test_data_2, #needs to be a matrix otherwise it imports as list
                    design = ~Region_Diet)

colData(dds) %>% head()
rowData(dds) %>% head()

counts(dds) %>% colnames()
dds
```








# 4. Function combined DESEQ2 
```{r}

f_DE_entire_run <- function(region, sample_to_remove, export_tables_T_F = T, use_IHW_filtering ){
coldata1 <- coldata  %>% 
  dplyr::filter(stringr::str_detect(rowname0, region, negate = F)) %>%
  dplyr::filter(!rowname0 %in% sample_to_remove) %>% 
  column_to_rownames(var = "rowname0")

cts0 <- cts %>%
  select(gene_ensembl, contains(region)) %>% 
  select(-any_of(sample_to_remove)) %>% 
  column_to_rownames(var = "gene_ensembl") %>%
  as.data.frame()


dds <<- DESeqDataSetFromMatrix(countData = cts0,
                              colData = coldata1,
                              design = ~ Region_Diet , 
                              tidy = F)

keep <- rowSums(counts(dds)) >= 10
paste("Number of genes with more than 10 counts accross all samples:  ", length(which(keep == T)))

dds <- dds[keep,]



#Run DESeq with Wald - pairwise

dds_wald <- DESeq(dds, test = "Wald", 
                 ) #this is default to n = 7; however this replace outlier gene values with a conservative estimate - this is very impor betaPrior = F, minReplicatesForReplace = 6tant for downstream PIF analysis. If n = <5/treatment it would be best to remove gene entirely.

#Check cooks distance boxplot to see if any gene is higher on average
#boxplot(log10(assays(dds_wald)[["cooks"]]), range=0, las=2)

boxplot_cooks_distance  <- log10(assays(dds_wald)[["cooks"]]) %>% as.data.frame() %>% 
  rownames_to_column("ensembl") %>% 
  pivot_longer(is.numeric, values_to = "cooks_distance") %>% 
ggplot(mapping = aes(x = cooks_distance, y = name))+
  geom_boxplot()+
  theme_classic()


compare_region <- region

if (use_IHW_filtering == TRUE) {
  f_IHW <- function(results_object) {IHW::ihw(results_object, alpha = alpha)}
}  else {
  f_IHW <- function(results_object){results_object}
}


#produce results
################################################################################## #
#HCP-HP-UMEI VS HCP-LP-UMEI - non-deficient vs low phos
################################################################################## #
res_HHUvsHLU <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","HCP.HP.UMEI"), 
                                               paste0(compare_region,"_","HCP.LP.UMEI")),
                        alpha = 0.05) %>% 
  f_IHW()

df_res_HHUvsHLU <- res_HHUvsHLU %>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input')  %>% arrange((padj)) 

message(crayon::green("HCP-HP-UMEI VS HCP-LP-UMEI - non-deficient vs low phos"))
summary(res_HHUvsHLU, alpha=0.05)


################################################################################## #
#HCP-HP-UMEI VS LCP-HP-UMEI - non-deficient vs low protein
################################################################################## #
res_HHUvsLHU <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","HCP.HP.UMEI"), 
                                               paste0(compare_region,"_","LCP.HP.UMEI")),
                        alpha = 0.05) %>% 
  f_IHW()

df_res_HHUvsLHU <- res_HHUvsLHU %>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input') %>% arrange((padj)) 

message(crayon::green("HCP-HP-UMEI VS LCP-HP-UMEI - non-deficient vs low protein"))
summary(res_HHUvsLHU, alpha=0.05)


################################################################################## #
#HCP-HP-UMEI VS LCP-LP-UMEI - non-deficient vs dual low
################################################################################## #
res_HHUvsLLU <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","HCP.HP.UMEI"), 
                                               paste0(compare_region,"_","LCP.LP.UMEI")),
                        alpha = 0.05) %>% 
  f_IHW()

df_res_HHUvsLLU <- res_HHUvsLLU %>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input')  %>% arrange((padj)) 

message(crayon::green("HCP-HP-UMEI VS LCP-LP-UMEI - non-deficient vs dual low"))
summary(res_HHUvsLLU, alpha=0.05)

################################################################################## #
#HCP-LP-UMEI VS LCP-HP-UMEI - low phos vs low protein
################################################################################## #
res_HLUvsLHU <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","HCP.LP.UMEI"), 
                                               paste0(compare_region,"_","LCP.HP.UMEI")),
                        alpha = 0.05) %>% 
  f_IHW()


df_res_HLUvsLHU <- res_HLUvsLHU%>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input')  %>% arrange((padj)) 

message(crayon::green("HCP-LP-UMEI VS LCP-HP-UMEI - low phos vs low protein"))
summary(res_HLUvsLHU, alpha=0.05)

################################################################################## #
#LCP-HP-UMEI VS LCP-LP-UMEI - low protein vs low dual low
################################################################################## #
res_LHUvsLLU <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","LCP.HP.UMEI"), 
                                               paste0(compare_region,"_","LCP.LP.UMEI")),
                        alpha = 0.05) %>% 
  f_IHW()


df_res_LHUvsLLU <- res_LHUvsLLU%>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input')  %>% arrange((padj)) 

message(crayon::green("LCP-HP-UMEI VS LCP-LP-UMEI - low protein vs low dual low"))
summary(res_LHUvsLLU, alpha=0.05)


################################################################################## #
#HCP-LP-UMEI VS LCP-LP-UMEI - low phos vs low dual low
################################################################################## #
res_HLUvsLLU <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","HCP.LP.UMEI"), 
                                               paste0(compare_region,"_","LCP.LP.UMEI")),
                        alpha = 0.05) %>% 
  f_IHW()


df_res_HLUvsLLU<- res_HLUvsLLU%>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input') %>% arrange((padj)) 

message(crayon::green("HCP-LP-UMEI VS LCP-LP-UMEI - low phos vs low dual low"))
summary(res_HLUvsLLU, alpha=0.05)


## RESTRICTED
################################################################################## #
#HCP-HP-UMEI VS HCP-HP-RMEI - non-deficient vs restricted
################################################################################## #
res_HHUvsHHR <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","HCP.HP.UMEI"), 
                                               paste0(compare_region,"_","HCP.HP.RMEI")),
                        alpha = 0.05) %>% 
  f_IHW()


df_res_HHUvsHHR <- res_HHUvsHHR %>% as.data.frame() %>% rownames_to_column("input") %>% 
 left_join(converted_genes, by = 'input') %>% arrange((padj)) 

message(crayon::green("HCP-HP-UMEI VS HCP-HP-RMEI - non-deficient vs restricted"))
summary(res_HHUvsHHR, alpha=0.05)

################################################################################## #
#HCP-LP-UMEI VS HCP-HP-RMEI - low phos vs restricted
################################################################################## #
res_HLUvsHHR <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","HCP.LP.UMEI"), 
                                               paste0(compare_region,"_","HCP.HP.RMEI")),
                        alpha = 0.05) %>% 
 f_IHW()

df_res_HLUvsHHR <- res_HLUvsHHR %>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input') %>% arrange((padj)) 

message(crayon::green("HCP-LP-UMEI VS HCP-HP-RMEI - low phos vs restricted"))
summary(res_HLUvsHHR, alpha=0.05)

################################################################################## #
#LCP-HP-UMEI VS HCP-HP-RMEI - low protein vs restricted
################################################################################## #
res_LHUvsHHR <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","LCP.HP.UMEI"), 
                                               paste0(compare_region,"_","HCP.HP.RMEI")),
                        alpha = 0.05) %>% 
  f_IHW()

df_res_LHUvsHHR <- res_LHUvsHHR %>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input')  %>% arrange((padj)) 

message(crayon::green("LCP-HP-UMEI VS HCP-HP-RMEI - low protein vs restricted"))
summary(res_LHUvsHHR, alpha=0.05)


################################################################################## #
#LCP-LP-UMEI VS HCP-HP-RMEI - dual low vs restricted
################################################################################## #
res_LLUvsHHR <- results(dds_wald, contrast = c("Region_Diet", 
                                               paste0(compare_region,"_","LCP.LP.UMEI"), 
                                               paste0(compare_region,"_","HCP.HP.RMEI")),
                        alpha = 0.05) %>% 
  f_IHW()

df_res_LLUvsHHR <- res_LLUvsHHR %>% as.data.frame() %>% rownames_to_column("input") %>% 
  left_join(converted_genes, by = 'input')  %>% arrange((padj)) 

message(crayon::green("LCP-LP-UMEI VS HCP-HP-RMEI - dual low vs restricted"))
summary(res_LLUvsHHR, alpha=0.05)


#import pairwise data for downstream
pairwise_contrasts <- fread("./pairwise_contrasts.csv") %>% 
  dplyr::rename(pairwise_short = df_pairwise_name) %>% 
  mutate(df_pairwise_name = str_c("df_res_",pairwise_short),
         res_pairwise_name = str_c("res_",pairwise_short))

 
#mutate descriptive columns to dataframes and create one long dataset of all results

full_dataset_DE <- plyr::ldply(pairwise_contrasts$df_pairwise_name, 
                        function(x) {
                          .pairwise_df <- dplyr::filter(pairwise_contrasts, df_pairwise_name == x)
                          #the following code could probably use ldply too? as it completes rbind.fill within
                          df <- do.call("rbind.fill",
                                        lapply(.pairwise_df$df_pairwise_name,
                                               function(x)
                                                 get(x)
                                                 ))#get's the dataframe by name
                          
                          df <- df %>% mutate(import_comparison_id = str_c(region, .pairwise_df$pairwise_short[1], sep = "_"),
                                              import_region = region,
                                              pairwise_short = .pairwise_df$pairwise_short[1],
                                              pairwise_description = .pairwise_df$Long_pairwise_description,
                                              pairwise_long = str_split(pairwise_description, pattern = " - ", simplify = T)[,1],
                                              description_short = str_split(description, pattern = "\\[", simplify = T)[,1]
                                              )
                          df
                          }
                        )


#iterates over all results objects to make plots
MA_pairwise_plots <- pbapply::pblapply(pairwise_contrasts$res_pairwise_name,
         function(x){
           res_data_to_plot <- get(x)
           
           DESeq2::plotMA(res_data_to_plot, main = paste(region, x))
           MA_plot <- recordPlot()
         
         })
#renames list elements to be a plot title
names(MA_pairwise_plots) <- str_c("MA_plot", region, pairwise_contrasts$pairwise_short, sep = "_")

#iterates over all results objects to make pvalue histograms (consider update with )
pvalue_histogram_pairwise_plots <- pbapply::pblapply(pairwise_contrasts$res_pairwise_name,
         function(x){
           res_data_to_plot <- get(x)
         
           plot.new()
           hist( res_data_to_plot$pvalue, breaks=50, col="grey", main = str_c("Histogram of ", region, " _ ", x))
           pvalue_histogram <- recordPlot()
      
        
         })
#renames list elements to be a plot title
names(pvalue_histogram_pairwise_plots) <- str_c("Pvalue_histogram", region, pairwise_contrasts$pairwise_short, sep = "_")

#add's plots into 1 named list for export
pairwise_plots <- list(MA_plots = MA_pairwise_plots, Pvalue_histogram = pvalue_histogram_pairwise_plots)
#export to global environment and append to list
pairwise_plots_out <<- c(pairwise_plots_out,setNames(list(pairwise_plots), paste0(region, "_pairwise_plots")))

#calculate normalised expression
vsd <- vst(dds, blind=FALSE) 
vsd_a <- assay(vsd) %>% as.data.frame() %>% rownames_to_column(var = "ensembl") %>% left_join(converted_genes, by = c("ensembl" = "input")) 
dds <- estimateSizeFactors(dds) #this is the 'factor' that it normalises by below

#test_data_a <- counts(dds, normalized=TRUE) %>% as.data.frame() %>% rownames_to_column(var = "ensembl") %>% left_join(converted_genes, by = c("ensembl" = "input"))

#to convert to log2, then add pseudocount of +1 then log2 it (bit rough)
#test_data_a <- data.frame(log2(counts(dds, normalized=TRUE)[, 1:ncol(dds)] + 1)) %>% rownames_to_column(var = "ensembl") %>% left_join(converted_genes, by = c("ensembl" = "input"))

#same as: 
log2norm <- normTransform(object = dds, 
                          f = log2,#log2 transformation
                          pc = 1) #pseudocount

# Annotate
log2norm_a <- assay(log2norm) %>% as.data.frame() %>% rownames_to_column(var = "ensembl") %>% left_join(converted_genes, by = c("ensembl" = "input")) 

new_coldata <- coldata1 %>% 
  rownames_to_column("rowname0") %>% 
    filter(Region %in% region) %>% 
    mutate(rowname2 = str_c(Region, "_", ID, "__", Diet))


vsd_new_names <- vsd_a %>%  dplyr::select(ensembl, name, description, new_coldata$rowname0)  #order columns in same order
names(vsd_new_names) <- c("ensembl","name", "description", new_coldata$rowname2) #rename columns with new treatment names included
head(vsd_new_names)

log2norm_new_names <- log2norm_a %>%  select(ensembl, name, description, new_coldata$rowname0)  #order columns in same order
names(log2norm_new_names) <- c("ensembl","name", "description", new_coldata$rowname2) #rename columns with new treatment names included
head(log2norm_new_names)

#export tables of VST and log2 normalised data, plus the full list of DE output
if(export_tables_T_F == TRUE){
fwrite(vsd_new_names , 
       file = paste0("./outputs/normalised_counts/",region,"_VST_normalisedcounts_",format(Sys.time(), "%Y%m%d_%H%M"),".csv"), row.names = F)

fwrite(log2norm_new_names, 
       file = paste0("./outputs/normalised_counts/",region,"_log2_of_DESEQ_internal_normalisedcounts_",format(Sys.time(), "%Y%m%d_%H%M"), ".txt"), row.names = F)

fwrite(full_dataset_DE,
       file = paste0("./outputs/",region,"_DE_output_", format(Sys.time(), "%Y%m%d_%H%M"), ".txt"))
}



#Create PCA plot and append to list
pcaData <- plotPCA(vsd, intgroup=c("Diet"), returnData=TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData, "percentVar"))
PCA_plot <- ggplot(pcaData, aes(PC1, PC2, fill=Diet)) +
  geom_point(size=4.2, shape = 21, colour = "black", alpha = 0.8) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()+
  #coord_fixed(xlim = c(-150,150), ylim = c(-120, 120)) +
  theme_classic()+
  geom_text(aes(label=name))+
  ggtitle(label = paste(region))



#Heatmap #************** Updated with sample distances pre-calculated with dist(t(assay(vsd))) - see vignette
topVarGenes <- head( order( rowVars( assay(vsd) ), decreasing=TRUE ), 500 )

heatmap_input <- assay(vsd)[ topVarGenes, ]
  
heatmap_plot<- pheatmap(heatmap_input, 
         scale = "row",
         #cutree_cols = 5,
         #cutree_rows = 4,
         clustering_method = "complete",
         #kmeans_k = 50,
         #annotation_row = 
         main = paste(region))


plots <- list( heatmap = heatmap_plot, PCA = print(PCA_plot), cooks_distance = boxplot_cooks_distance)


Plot_out_list <<- c(Plot_out_list, setNames(list(plots), paste0(region, "_plots")))

#append normalised data and DE_out to list for export
data_out <-  list(vsd = vsd_new_names, log2norm = log2norm_new_names, DE_out = full_dataset_DE)

Norm_and_DESEQ_out_list <<- c(Norm_and_DESEQ_out_list, setNames(list(data_out), paste0(region, "_data")))

}


```
#Working
Subset se object before converting dds (as dds should be in function as split by region.)



a character vector with exactly three elements: the name of a factor in the design formula, the name of the numerator level for the fold change, and the name of the denominator level for the fold change (simplest case)
```{r}

keep <- rowSums(counts(dds)) >= 10
paste("Number of genes with more than 10 counts accross all samples:  ", length(which(keep == T)))

dds <- dds[keep,]



#Run DESeq with Wald - pairwise
#testin on LIV only
dds_wald <- DESeq(dds, test = "Wald", 
                 )



#The contrasts are built using the different contrasts available in Region_diet



design(dds_wald)

colData(dds_wald)


model.matrix(design(dds_wald), colData(dds_wald))

mcols(dds_wald)

resultsNames(dds_wald)

###############################################################################################################


comparisons <- c("HHUvsLLU", "HHUvsHHR")







dds_object <- dds_wald

top_level_name <- "LIV"
contrast_factor <- "Region_Diet"



if (use_IHW_filtering == TRUE) {
  f_IHW <- function(results_object) {IHW::ihw(results_object, alpha = alpha)}
}  else {
  f_IHW <- function(results_object){results_object}
}



f_calculate_DESEQ_results <- 
function(dds_object, 
         contrast_factor,  #Name of column containing treatments to contrast
         combinations = NA, #If NA, uses make_pairwise_combinations(); should be a list with each element of list 2 strings long
         top_level_name, #string for top level, e.g. Region
         alpha = 0.05){
  
  test <<- colData(dds_object)
  
  if(is.na(combinations)){
    combinations <- 
      make_pairwise_combinations(SummarizedExperiment::colData(dds_object), 
                                 contrast_factor = contrast_factor, 
                                 top_level_name = top_level_name)
  } 
  
  
  f_make_results <- 
    function(x, dds_object, contrast_factor, alpha = alpha){
    contrast_numerator <- x[1]
    contrast_denominator <- x[2]
    
    res <- results(dds_object, 
            contrast = c(contrast_factor,
                         contrast_numerator,
                         contrast_denominator),
            alpha = alpha) %>%   f_IHW()
    
    message(crayon::green(paste("\n",res@elementMetadata$description[2])))
    summary(res, alpha=alpha)
    res
  }
  
  res_out <- 
    pbapply::pblapply(combinations, #iterate over all combinations
                      f_make_results, #using this internal function
                      dds_object = dds_object,
                      contrast_factor = contrast_factor, 
                      alpha = alpha)
  
  #set names of the res_out list
   res_out <- setNames(res_out, 
                       lapply(combinations, 
                              function(x) {paste(x, collapse = " vs ")})
                       )
 
  
  #res_out is a list of all res objects for each comparison, named.
  
   res_out

}


results_out_test <-
  f_calculate_DESEQ_results(dds_object = dds_wald,
                          contrast_factor = "Region_Diet",
                          combinations = NA ,
                          top_level_name = "Region",
                          alpha = 0.05)




#to access results as a dataframe

lapply(res_out, function(x){as.data.frame(x) %>% 
    rownames_to_column("input") %>% 
    #left_join %>% 
    arrange(padj)
  })


```

