---
title: "Testing DESEQ2"
output: html_notebook
---

Iteration of DESEQ2 over: Region level, then treatment comparison

Region level - e.g. Tissue region, indicates all DE data from same tissue and experiment. These are analysed independently from every other region.

Within a region, the treatments are compared.


However, for downstream plotting, the normalised data from all regions is calculated. Particularly important for co-expression related analysis.


To normalise gene expression across all tissues, the 'DESEQ2 - all tissues - FULL - **' needs to be used.

#Step 1a in pipeline

# 1. Import metadata
```{r}

#import data about columns, that is each column should be metadata about the sample/animal. Particularly include all treatment info and animal ID in columns. 
coldata <-  data.table::fread(file = "../Sheep_RNAseq/Inputs/coldata_new_names_20201105.csv") %>% 
  dplyr::mutate(sample_names = paste0(Region, "_", ID, "__", Diet))

coldata <-  data.table::fread(file = "./coldata_test.csv") %>% 
  dplyr::mutate(sample_names = paste0(Region, "_", ID, "__", Diet))


```




# 2. import all count tables and join
TODO - make it handle when too many names are in coldata compared to cts_all
Error: Can't subset columns that don't exist.
x Columns `BONE_8__Feed`, `BONE_20__Fast`, and `BONE_15__Refeed` don't exist.
```{r matrix_input, echo=T}


cts_all <- data.table::fread(file= "./cts_renamed_all_Sheep.txt")

###### ************
# Use new function
cts_filtered <- check_count_matrix(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names
                            )

# filter colData, as suggested by above
coldata <- subset_colData(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names)

#re-run check
cts_filtered <- check_count_matrix(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names
                            )
#rm(cts_all)
```



# 3. convert ensemble ID to gene ID

```{r}

test_annot <- annotate_gene_ensembl(cts_filtered, organism = "oaries")
```


# Prepare data
TO DO: make a wrapper for this with specific checks of data / auto matrix
```{r}

test_data <-
  SummarizedExperiment::SummarizedExperiment(
    assays = cts_filtered %>% 
      tibble::column_to_rownames(var = "gene_ensembl") %>% 
      as.matrix(),
    rowData = test_annot, 
    colData = coldata)

#subset columns to be every third, to make testing quicker
test_data <- test_data[,seq(0,ncol(cts_filtered)-1,3)] 


```



#4. Generate DE results
TO DO: - detect all options for 'top_level_filter' and parse to map().

fails due to ihw error: Error in mcols(deseq_res) : could not find function "mcols"
```{r}

DE_out <- 
  purrr::map(list("LIV","ARC"),
             auto_generate_DE_results, 
             se_data = test_data,
             top_level_name = Region,
             column_of_samples = sample_names,
             samples_to_remove = NA,
             DESeq2_formula_design = ~Region_Diet,
             rowSums_filter = 10, #for dds filtering
             results_contrast_factor = Region_Diet,
             results_combinations = NA,
             use_IHW_filtering = TRUE,
             alpha = 0.05,
             gene_annotations = test_annot,
             export_tables = TRUE,
             export_dir = "./outputs/normalised_counts3/")



```

#5. Accessing results
TO DO: 
```{r}
#unlist to take away teh unamed, top layer. Gives access via $
DE_out_test <- DE_out %>% unlist(recursive = FALSE)

#{.[grepl("MA_plots", names(.))]}
```



# Code for renaming old cts tables
```{r eval=FALSE, include=FALSE}
#read in count table - hypothalamus

cts_hypo_liv <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_hypothalamus-liver.txt", header = T)

colnames(cts_hypo_liv)[1] <- "gene_ensembl"

#rename incorrect entries:
cts_hypo_liv <- cts_hypo_liv %>% 
  dplyr::rename("VMH_21_HCP-HP-UMEI" = "VMH_21_HCP-HP-RMEI", 
                "VMH_02_HCP-LP-UMEI" = "VMH_02_LCP-LP-UMEI")


#rename LAT to LHA
colnames(cts_hypo_liv) <- stringr::str_replace_all(string = names(cts_hypo_liv), pattern = "LAT", replacement = "LHA")

#find actual sample names from table
selection <- match(colnames(cts_hypo_liv)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)

new_names <- coldata[selection,]$sample_names
colnames(cts_hypo_liv)[-1] <- new_names

selection <- match(colnames(cts_hypo_liv)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)
all((temp1[["rowname"]]== colnames(cts_hypo_liv)[-1] ))


#read in count table - GIT & ST
cts_git_st <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_matrix_DUO_RUM_ST_20200901.tabular") %>% 
  dplyr::select(!contains("ST"))

colnames(cts_git_st)[1] <- "gene_ensembl"


selection <- match(colnames(cts_git_st)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)
temp1 <- coldata[selection,]
all(temp1[["rowname"]] == colnames(cts_git_st)[-1] )

new_names <- temp1$sample_names
colnames(cts_git_st)[-1] <- new_names
#Join cts together

cts_all <- dplyr::full_join(cts_hypo_liv, cts_git_st)

colnames(cts_all)
```

