---
title: "DESEQ2 - Initial"
output: html_notebook
---

Iteration of DESEQ2 over: Top_level_filter (e.g. Tissue Region) level, then treatment comparison

Region level - e.g. Tissue region, indicates all DE data from same tissue and experiment. These are analysed independently from every other region.

Within a region, the treatments are compared.


However, for downstream plotting, the normalised data from all regions is calculated. Particularly important for co-expression related analysis.


To normalise gene expression across all tissues, the 'DESEQ2 - all tissues - FULL - **' needs to be used.


# 1. Import metadata
```{r}

#import data about columns, that is each column should be metadata about the sample/animal. Particularly include all treatment info and animal ID in columns. 
 coldata <-  data.table::fread(file = "./extdata/coldata_new_names_20201105.csv") %>% 
   dplyr::mutate(sample_names = paste0(Region, "_", ID, "__", Diet))

#test to throw error
#coldata <-  data.table::fread(file = "./coldata_test.csv") %>% 
 # dplyr::mutate(sample_names = paste0(Region, "_", ID, "__", Diet))



```




# 2. import all count tables and join
```{r matrix_input, echo=T}
cts_all <- data.table::fread(file= "./extdata/cts_renamed_all_Sheep.txt")

# 1. Run check function
cts_filtered <- check_count_matrix(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names
                            )

# 2. filter colData, if suggested by above (will work if not suggested anyway)
coldata <- subset_colData(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names)

# 3. re-run check
cts_filtered <- check_count_matrix(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names
                            )
rm(cts_all)
```



# 3. convert ensemble ID to gene ID

```{r}

test_annot <- 
  annotate_gene_ensembl(cts_filtered, 
                        organism = "oaries",
                        base_URL = "http://biit.cs.ut.ee/gprofiler_archive3/e100_eg47_p14")


#used without internet connection
# temp <- data.table::fread(file="./outputs/normalised_counts/ARC_VST_normalisedcounts_20211117_1638.txt")
# test_annot <- temp %>% dplyr::select(gene_ensembl, gene_name, description)
# test_annot <- cts_filtered %>% 
#   dplyr::select(gene_ensembl) %>% 
# dplyr::left_join(test_annot)
```


# Prepare data
TO DO: make a wrapper for this with specific checks of data / auto matrix
```{r}


test_data <- 
make_summarized_experiment_object(counts_data = cts_filtered,
                                  gene_annotations = test_annot, 
                                  colData = coldata)



#subset columns to be every n, to make testing quicker
#test_data <- test_data[,seq(0,ncol(cts_filtered)-1,2)] 


```



# 4. Generate DE results
The function `auto_generate_DE_results()` is iterated over using `purrr::map()`.
For details on variables, see `?auto_generate_DE_results`.
For multiple iterations it can take a long time to run. E.g. 6 Regions x 5 treatments of 8 samples (total n = 240) = ~30mins
```{r}

# regions <- coldata$Region %>% unique()

DE_out <- 
  auto_generate_DE_results(se_data = test_data,
                           top_level_groups = c("ARC", "LIV"),
                           top_level_colname = Region,
                           sample_colname = sample_names,
                           samples_to_remove = NA,
                           DESeq2_formula_design = ~Region_Diet,
                           rowSums_filter = 10, #for dds filtering
                           results_contrast_factor = Region_Diet,
                           results_combinations = NA,
                           use_IHW_filtering = TRUE,
                           alpha = 0.05,
                           gene_annotations = test_annot,
                           export_tables = FALSE,
                           export_dir = "./outputs/",
                           whole_data_normalisation = FALSE)



```

# 5. Save Results


Saving to .rds compresses all of the data into one file that can be re-imported into RStudio, without having to re-run the very long commands above. It does not re-instate the whole image of the environment, just the output of auto_generate_DE_results() which has been unlisted so that every entry is named.

***The call to saveRDS() can take ~ 10 minutes and still produces a large file, particularly if many regions and samples are included. E.g. 40 samples x 6 regions = ~2.0 GB 
```{r eval=FALSE, include=TRUE}
# Defaults to NOT RUN when running whole document.
#Save file as progress. Can be used as input instead of re-running all of the DESEQ functions.
saveRDS(DE_out, file = "./outputs/DE_out_file_testing.rds")

#to re-import to environment:
#DE_out2 <- readRDS(file = "./outputs/DE_out_file.rds")

```

# 6. Accessing Results

## 6.a. Accessor Functions
Firstly, the use of accessor functions that help to extract specific data are shown
below. For plots, these also 'view' the plots, but is generally only useful for
a code chunk that can show all plots, such as in this document 
(rather than if using in a .R file format). Note that the size of these plots is
dictated by the `fig.width = ` in the `{r }` at beginning of each code chunk. 

These functions return a named list of 1 level. This means that individual 
elements of the list can be accessed by using $. For some plot types, viewing the 
individual plot using the $ will need to be done by copying to the Console. Do this
if the plot does not appear below code chunk.

A description of more advanced access techniques is shown in next section.

### PIF Tables
All genes, sorted by PIF value.
This function also allows searching to return only specified groups. 
```{r}
PIF_tables_ALL <- GET_PIF_df(auto_DE_output = DE_out)

PIF_tables_ARC_and_LIV <- GET_PIF_df(auto_DE_output = DE_out,
                                  search_top_level = c("ARC", "LIV"))

rm(PIF_tables_ALL)
rm(PIF_tables_ARC_and_LIV)
```

### DE tables, sorted by PIF
Showing only significant genes, sorted on absolute PIF value.
This function also allows searching to return only specified groups. 
```{r}
DE_tables_ALL <- GET_DE_by_PIF_df(auto_DE_output = DE_out)

DE_tables_ARC_and_LIV <- GET_DE_by_PIF_df(auto_DE_output = DE_out,
                                          search_top_level = c("ARC", "LIV"))

rm(DE_tables_ALL)
rm(DE_tables_ARC_and_LIV)
```

#### Individual normalised counts tables
Note - these are largely not useful. See ?GET_normalised_counts
Searches are not required, and either or both searches can be used.
```{r}
norm_ARC_log2 <- GET_normalised_data(auto_DE_output = DE_out, 
                                     search_top_level = "ARC",
                                     vst_or_log2 = "Log2")
```

### Overall Plots

#### Boxplot of cooks distances
This has own function compared to others below. This is useful for both DE 
comparisons and when whole data normalisation has been run.
```{r}
pairwise_boxplots <- GET_boxplot_cooks_distance(auto_DE_output = DE_out)
```

#### PCA
```{r fig.width = 8}
PCA_plots <- GET_overall_plots(auto_DE_output = DE_out,
                   type = "PCA")

#PCA_plots$ARC_DESeq2_Output
```

#### Heatmap
```{r fig.width = 8}
heatmap_plots <- GET_overall_plots(auto_DE_output = DE_out,
                   type = "heatmap")

```

#### Sample distances
```{r fig.width = 8}
sample_dist_plots <- GET_overall_plots(auto_DE_output = DE_out,
                   type = "sample_distances")
```

### Pairwise plots

#### Pvalue Histrograms
```{r fig.width = 10}

pvalue_histograms <- GET_pairwise_plots(auto_DE_output = DE_out,
                                        type = "Pvalue", 
                                        search_top_level = c("ARC", "LIV"))

```



#### MA Plots (for QC)
These are not the same MA plots which are generated with PIF, DE and annotations. This is rough plots from DESeq2.
```{r fig.width = 10}

MA_plots <- GET_pairwise_plots(auto_DE_output = DE_out,
                               type = "MA",
                               search_top_level = c("ARC"))

```

### DESeqDataSet (dds_wald) objects
These are the DESeqDataSet objects (derived from the SummarizedExperiment object parsed to `auto_generate_DE_results`) after the call to DESeq2 (via `auto_generate_DE_results`).
```{r}
dds_objects <- GET_dds_wald_objects(auto_DE_output = DE_out,
                     search_top_level = "LIV")
#rm(dds_objects)
```


### Get results objects (class DESeqResults)
These are the results objects, containing all metadata, produced for 
each pairwise comparison. Most useful for accessing advanced information on how
results were generated. 
```{r}
results_objects <- GET_DESeq2_res_object(auto_DE_output = DE_out,
                                         search_top_level = c("ARC", "LIV"))

#rm(results_objects)
```





## 6.b. Manually accessing results
Overall, results can be accessed by using $ notation to navigate through the list.
Search, unlisting and reduction techniques help to select data of interest.

The first level of the list can be accessed by using the $ sign. To see all available options use names().

```{r}
#DE_out$ARC_DESeq2_Output

DE_out %>% names()
```

Multiple $ can be used for recurring levels. Pressing the Tab button on keyboard shows options as well.
Likewise, if a dataframe is the lowest level of the list that is accessed with a $, then it will also allow the column names to be accessed with \$ as well. 
```{r eval=FALSE}
#Access a dataframe
DE_out$ARC_DESeq2_Output$DESeq2_annot_df$`ARC_HCP.HP.RMEI vs ARC_HCP.HP.UMEI`

#Access a column in a dataframe
DE_out$ARC_DESeq2_Output$DESeq2_annot_df$`ARC_HCP.HP.RMEI vs ARC_HCP.HP.UMEI`$gene_ensembl
```

Using a call to purrr::map with a string will iterate over each of the entries in the first level of the list and search within those lists to return a new list of only entries with matching names. This is basically what the \code{GET_*} functions use to access results.
```{r}
#Check what second level entries exist 
DE_out$ARC_DESeq2_Output %>% names()

```

```{r}
#The string in call to map must be one of the options produced from above. 
#This works because each Region's entry has the same underlying structure.
sample_accessed_object <-  DE_out %>% 
  purrr::map("dds_wald_object") 

sample_accessed_object %>%  str(max.level = 2)
```

A call to unlist(recursive = FALSE) removes the first layer of the list, which is often helpful if it returns a further list. E.g. there are 60 results objects when there are 10 pairwise comparisons x 6 regions.
```{r}

DE_out %>% 
  purrr::map("DESeq2_res_object") %>% 
  unlist(recursive = FALSE) %>% 
  str(max.level = 2, list.len = 10) #shows the structure of the top 2 levels of the first 10 of 60 entries for easier reading
```




# 7. Filter and re-run
After viewing the results, particularly the cooks distances and the overall plots, a decision can be made to exclude outlier samples. For example, one might choose to remove the samples named: "ARC_88__LCP-LP-UMEI" and "LIV_77__HCP-HP-UMEI"
This can be specified with an additional call to `auto_generate_DE_results`
The same GET_* functions can then be used to access results from DE_out_2.

```{r}
#call DE again, without these samples
DE_out_2 <- 
  auto_generate_DE_results(se_data = test_data$sample_names,
                           top_level_groups = c("ARC", "LIV"),
                           top_level_colname = Region,
                           sample_colname = sample_names,
                           samples_to_remove = c("ARC_88__LCP-LP-UMEI", 
                                                 "LIV_77__HCP-HP-UMEI"),
                           DESeq2_formula_design = ~Region_Diet,
                           rowSums_filter = 10, #for dds filtering
                           results_contrast_factor = Region_Diet,
                           results_combinations = NA,
                           use_IHW_filtering = TRUE,
                           alpha = 0.05,
                           gene_annotations = test_annot,
                           export_tables = FALSE,
                           export_dir = "./outputs/",
                           whole_data_normalisation = FALSE)

```


# 8. Generate whole dataset normalised counts
Remember to include any samples to remove, if required.
```{r}

full_norm <- 
  auto_generate_DE_results(se_data = test_data,
                           top_level_colname = Region,
                           sample_colname = sample_names,
                           samples_to_remove = NA,
                           DESeq2_formula_design = ~Region_Diet,
                           gene_annotations = test_annot,
                           export_tables = FALSE,
                           export_dir = "./outputs/",
                           whole_data_normalisation = TRUE) #This is important!


```

# boxplot cooks distance
```{r}
#full_norm$Whole_data_normalisation_output$boxplot_cooks_distance
GET_boxplot_cooks_distance(full_norm)
```


#Working
Code for renaming old cts tables
```{r eval=FALSE, include=FALSE}
#read in count table - hypothalamus

cts_hypo_liv <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_hypothalamus-liver.txt", header = T)

colnames(cts_hypo_liv)[1] <- "gene_ensembl"

#rename incorrect entries:
cts_hypo_liv <- cts_hypo_liv %>% 
  dplyr::rename("VMH_21_HCP-HP-UMEI" = "VMH_21_HCP-HP-RMEI", 
                "VMH_02_HCP-LP-UMEI" = "VMH_02_LCP-LP-UMEI")


#rename LAT to LHA
colnames(cts_hypo_liv) <- stringr::str_replace_all(string = names(cts_hypo_liv), pattern = "LAT", replacement = "LHA")

#find actual sample names from table
selection <- match(colnames(cts_hypo_liv)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)

new_names <- coldata[selection,]$sample_names
colnames(cts_hypo_liv)[-1] <- new_names

selection <- match(colnames(cts_hypo_liv)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)
all((temp1[["rowname"]]== colnames(cts_hypo_liv)[-1] ))


#read in count table - GIT & ST
cts_git_st <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_matrix_DUO_RUM_ST_20200901.tabular") %>% 
  dplyr::select(!contains("ST"))

colnames(cts_git_st)[1] <- "gene_ensembl"


selection <- match(colnames(cts_git_st)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)
temp1 <- coldata[selection,]
all(temp1[["rowname"]] == colnames(cts_git_st)[-1] )

new_names <- temp1$sample_names
colnames(cts_git_st)[-1] <- new_names
#Join cts together

cts_all <- dplyr::full_join(cts_hypo_liv, cts_git_st)

colnames(cts_all)
```

