---
title: "DESEQ2 - Initial"
output: html_notebook
---

Iteration of DESEQ2 over: Top_level_filter (e.g. Tissue Region) level, then treatment comparison

Region level - e.g. Tissue region, indicates all DE data from same tissue and experiment. These are analysed independently from every other region.

Within a region, the treatments are compared.


However, for downstream plotting, the normalised data from all regions is calculated. Particularly important for co-expression related analysis.


To normalise gene expression across all tissues, the 'DESEQ2 - all tissues - FULL - **' needs to be used.


# 1. Import metadata
```{r}

#import data about columns, that is each column should be metadata about the sample/animal. Particularly include all treatment info and animal ID in columns. 
# coldata <-  data.table::fread(file = "../Sheep_RNAseq/Inputs/coldata_new_names_20201105.csv") %>% 
#   dplyr::mutate(sample_names = paste0(Region, "_", ID, "__", Diet))

#test to throw error
#coldata <-  data.table::fread(file = "./coldata_test.csv") %>% 
 # dplyr::mutate(sample_names = paste0(Region, "_", ID, "__", Diet))

#loaded via package
data(coldata)

```




# 2. import all count tables and join
```{r matrix_input, echo=T}
#loaded via package
data(cts_all)

# cts_all <- data.table::fread(file= "./cts_renamed_all_Sheep.txt")

# 1. Run check function
cts_filtered <- check_count_matrix(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names
                            )

# 2. filter colData, if suggested by above (will work if not suggested anyway)
coldata <- subset_colData(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names)

# 3. re-run check
cts_filtered <- check_count_matrix(count_data = cts_all, 
                            colData = coldata,
                            column_with_col_names = sample_names
                            )
rm(cts_all)
```



# 3. convert ensemble ID to gene ID

```{r}

test_annot <- 
  annotate_gene_ensembl(cts_filtered, 
                        organism = "oaries",
                        base_URL = "http://biit.cs.ut.ee/gprofiler_archive3/e100_eg47_p14")


#used without internet connection
# temp <- data.table::fread(file="./outputs/normalised_counts/ARC_VST_normalisedcounts_20211117_1638.txt")
# test_annot <- temp %>% dplyr::select(gene_ensembl, gene_name, description)
# test_annot <- cts_filtered %>% 
#   dplyr::select(gene_ensembl) %>% 
# dplyr::left_join(test_annot)
```


# Prepare data
TO DO: make a wrapper for this with specific checks of data / auto matrix
```{r}

test_data <-
  SummarizedExperiment::SummarizedExperiment(
    assays = cts_filtered %>% 
      tibble::column_to_rownames(var = "gene_ensembl") %>% 
      as.matrix(),
    rowData = test_annot, 
    colData = coldata)

#subset columns to be every n, to make testing quicker
test_data <- test_data[,seq(0,ncol(cts_filtered)-1,2)] 


```



# 4. Generate DE results
The function `auto_generate_DE_results()` is iterated over using `purrr::map()`.
For details on variables, see `?auto_generate_DE_results`.
For multiple iterations it can take a long time to run. E.g. 6 Regions x 5 treatments of 8 samples (total n = 240) = ~30mins
```{r}
#get all top_level_filter options (which is parsed as .x in call to map; see ?auto_generate_DE_results)
regions <- test_data$Region %>% unique() %>% as.list()


DE_out1 <- 
  purrr::map(.x = regions[4:5],
             .f = auto_generate_DE_results, 
             se_data = test_data,
             top_level_name = Region,
             column_of_samples = sample_names,
             samples_to_remove = NA,
             DESeq2_formula_design = ~Region_Diet,
             rowSums_filter = 10, #for dds filtering
             results_contrast_factor = Region_Diet,
             results_combinations = NA,
             use_IHW_filtering = TRUE,
             alpha = 0.05,
             gene_annotations = test_annot,
             export_tables = TRUE,
             export_dir = "./outputs/",
             whole_data_normalisation = FALSE)

#DE_out %>% str(max.level = 3)
#unlist to take away the un-named, top layer. Gives access via $
DE_out <- DE_out1 %>% unlist(recursive = FALSE)

```

# 5. Save Results


Saving to .rds compresses all of the data into one file that can be re-imported into RStudio, without having to re-run the very long commands above. It does not re-instate the whole image of the environment, just the output of auto_generate_DE_results() which has been unlisted so that every entry is named.

***The call to saveRDS() can take ~ 10 minutes and still produces a large file, particularly if many regions and samples are included. E.g. 40 samples x 6 regions = ~2.0 GB 
```{r eval=FALSE, include=TRUE}
# Defaults to NOT RUN when running whole document.
#Save file as progress. Can be used as input instead of re-running all of the DESEQ functions.
saveRDS(DE_out, file = "./outputs/DE_out_file_testing.rds")

#to re-import to environment:
#DE_out2 <- readRDS(file = "./outputs/DE_out_file.rds")

```

# 6. Accessing Results
Overall, results can be accessed by using $ notation to navigate through the list.
Search, unlisting and reduction techniques help to select data of interest.

The first level of the list can be accessed by using the $ sign. To see all available options use names().

```{r}
#DE_out$ARC_DESeq2_Output

DE_out %>% names()
```
Multiple $ can be used for recurring levels. Pressing the Tab button on keyboard shows options as well.
Likewise, if a dataframe is the lowest level of the list that is accessed with a $, then it will also allow the column names to be accessed with \$ as well. 
```{r eval=FALSE}
#Access a dataframe
DE_out$ARC_DESeq2_Output$DESeq2_annot_df$`ARC_HCP.HP.RMEI vs ARC_HCP.HP.UMEI`

#Access a column in a dataframe
DE_out$ARC_DESeq2_Output$DESeq2_annot_df$`ARC_HCP.HP.RMEI vs ARC_HCP.HP.UMEI`$gene_ensembl
```

Using a call to purrr::map with a string will iterate over each of the entries in the first level of the list and search within those lists to return a new list of only entries with matching names. 
```{r}
#Check what second level entries exist 
DE_out$ARC_DESeq2_Output %>% names()

```
```{r}
#The string in call to map must be one of the options produced from above. 
#This works because each Region's entry has the same underlying structure.
sample_accessed_object <-  DE_out %>% 
  purrr::map("dds_wald_object") 

sample_accessed_object %>%  str(max.level = 2)
```

A call to unlist(recursive = FALSE) removes the first layer of the list, which is often helpful if it returns a further list. E.g. there are 60 results objects when there are 10 pairwise comparisons x 6 regions.
```{r}

DE_out %>% 
  purrr::map("DESeq2_res_object") %>% 
  unlist(recursive = FALSE) %>% 
  str(max.level = 2, list.len = 10) #shows first 10 of 60 for easier reading
```


## DE
These results are exported to individual worksheets in an excel file if export_tables = TRUE was selected in original run. However, they can also be accessed within R using the following techniques.
### Dataframes / tables
#### DE tables
These tables include only the genes <= alpha (which is default 0.05),sorted on absolute PIF value
```{r}
#Select DE tables to a list with 1 layer
DE_dataframes <- 
  DE_out %>% 
  purrr::map("DE_sig_PIF_df") %>% 
  unlist(recursive = FALSE)

#access individual tables by using $ 
#e.g.
DE_dataframes %>% str(list.len = 4)

#Can also search with a string e.g. "ARC"
ARC_DE  <- DE_dataframes %>% 
  {.[grepl("ARC", names(.))]} 

ARC_DE$`ARC_DESeq2_Output.ARC_HCP.HP.RMEI vs ARC_LCP.LP.UMEI` %>% View()

```

#### PIF tables
```{r}
PIF_dataframes <- 
  DE_out %>% 
  purrr::map("PIF") %>%   
  unlist(recursive = FALSE)

  
 #access individual tables by using $ 
#e.g.
PIF_dataframes$`ARC_DESeq2_Output.PIF - ARC_HCP.HP.UMEI vs ARC_LCP.LP.UMEI` %>% head()
```


### Results objects
Same technique applies as above
```{r}
#Select DE tables to a list with 1 layer
DE_results_objects <- 
  DE_out %>% 
  purrr::map("DESeq2_res_object") %>% 
  unlist(recursive = FALSE)

#access individual tables by using $ 
#e.g.
DE_results_objects$`LIV_DESeq2_Output.LIV_HCP.HP.UMEI vs LIV_LCP.LP.UMEI`

```
## Stored plots
### Pairwise plots
```{r}
#View which pairwise plots are available to access
DE_out$ARC_DESeq2_Output$pairwise_plots %>% names()
```

#### Pvalue Histrograms
Some plots are stored in different formats, and might require a small function to view them.
Note that the plots are visualised below this code chunk. The width of the figure can be adjusted with the `fig.width = ` in the first line of code chunk.
```{r fig.width = 10}

#Store pvalue histograms into a new object
pvalue_plots_list <- 
  DE_out %>% 
  purrr::map("pairwise_plots") %>%  
  purrr::map('Pvalue_histogram') %>% 
  unlist(recursive = FALSE)
  
#View plots (Returned in output below this code chunk)

  pvalue_plots_list[1:20] %>%  #showing first 20 only
  purrr::map(function(x){
    print(x)
    plot.new()
    dev.off()
    })
```
#### MA Plots (for QC)
These are not the same MA plots which are generated with PIF, DE and annotations. This is rough plots from DESeq2.
```{r fig.width = 10}

#Store pvalue histograms into a new object
MA_plots_list <- 
  DE_out %>% 
  purrr::map("pairwise_plots") %>%  
  purrr::map('MA_plots') %>% 
  unlist(recursive = FALSE)
  

#View plots (Returned in output below this code chunk)

  MA_plots_list[10:20] %>%  #showing 10 only
  purrr::map(function(x){
    print(x)
    plot.new()
    dev.off()
    })
```

### Overall Plots
```{r}
#View which overall plots are available to access
DE_out$ARC_DESeq2_Output$overall_plots %>% names()
```

#### Boxplot of cooks distances
```{r}

```




#### PCA
Some plots are in a different format (ggplots) and will show without function. e.g. the PCA plots
```{r}
#Store pvalue histograms into a new object
PCA_list <- 
  DE_out %>% 
  purrr::map("overall_plots") %>%  
  purrr::map('PCA') 

#View plots (Returned in output below this code chunk)
PCA_list
  
```

#### Heatmaps
```{r fig.width = 10}
#Store pvalue histograms into a new object
heatmaps <- 
  DE_out %>% 
  purrr::map("overall_plots") %>%  
  purrr::map('heatmap') 
  
#View plots (Returned in output below this code chunk)

  heatmaps %>% 
  purrr::map(function(x){
    print(x)
    plot.new()
    dev.off()
    })
```

#### Sample-Sample Distances
```{r fig.width = 10}
#Store pvalue histograms into a new object
samp_dist_list <- 
  DE_out %>% 
  purrr::map("overall_plots") %>%  
  purrr::map('Sample_distances') 
  
#View plots (Returned in output below this code chunk)

  samp_dist_list %>% 
  purrr::map(function(x){
    print(x)
    plot.new()
    dev.off()
    })
```


# 7. Filter and re-run
```{r}

```


# 8. Generate whole dataset normalised counts
```{r}
full_norm <- 
  auto_generate_DE_results(se_data = test_data,
                           top_level_name = Region,
                           column_of_samples = sample_names,
                           samples_to_remove = NA,
                           DESeq2_formula_design = ~Region_Diet,
                           gene_annotations = test_annot,
                           export_tables = TRUE,
                           export_dir = "./outputs/",
                           whole_data_normalisation = TRUE)


```

```{r}
full_norm$Whole_data_normalisation_output$normalised_data$vsd
```




#Workign
```{r eval = FALSE}
#testing how many have real names
 x %>% 
        as.data.frame() %>% 
        tibble::rownames_to_column(var = "gene_ensembl") %>%
        dplyr::left_join(annot, by = c("gene_ensembl")) %>%
        dplyr::select(.data$gene_ensembl,
                      .data$gene_name,
                      .data$description,
                      tidyselect::everything()) %>% 
        arrange(.data$padj)
```

## Code for renaming old cts tables
```{r eval=FALSE, include=FALSE}
#read in count table - hypothalamus

cts_hypo_liv <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_hypothalamus-liver.txt", header = T)

colnames(cts_hypo_liv)[1] <- "gene_ensembl"

#rename incorrect entries:
cts_hypo_liv <- cts_hypo_liv %>% 
  dplyr::rename("VMH_21_HCP-HP-UMEI" = "VMH_21_HCP-HP-RMEI", 
                "VMH_02_HCP-LP-UMEI" = "VMH_02_LCP-LP-UMEI")


#rename LAT to LHA
colnames(cts_hypo_liv) <- stringr::str_replace_all(string = names(cts_hypo_liv), pattern = "LAT", replacement = "LHA")

#find actual sample names from table
selection <- match(colnames(cts_hypo_liv)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)

new_names <- coldata[selection,]$sample_names
colnames(cts_hypo_liv)[-1] <- new_names

selection <- match(colnames(cts_hypo_liv)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)
all((temp1[["rowname"]]== colnames(cts_hypo_liv)[-1] ))


#read in count table - GIT & ST
cts_git_st <- data.table::fread("../Sheep_RNAseq/Inputs/featureCounts_matrix_DUO_RUM_ST_20200901.tabular") %>% 
  dplyr::select(!contains("ST"))

colnames(cts_git_st)[1] <- "gene_ensembl"


selection <- match(colnames(cts_git_st)[-1], coldata[["rowname"]] ) %>% purrr::discard(is.na)
temp1 <- coldata[selection,]
all(temp1[["rowname"]] == colnames(cts_git_st)[-1] )

new_names <- temp1$sample_names
colnames(cts_git_st)[-1] <- new_names
#Join cts together

cts_all <- dplyr::full_join(cts_hypo_liv, cts_git_st)

colnames(cts_all)
```

